io/include/pcl/io/openni_camera/openni_driver.h

```diff
@@ -52,201 +52,204 @@
+   * @return the connection string of the device.
+   */
+  const char* getConnectionString (unsigned index) const throw ();
+
+  /**
+   * @author Suat Gedikli
+   * @brief method to retrieve the vendor name of the USB device without creating it.
+   * @param[in] index the index of the device in the device list.
+   * @return the vendor name of the USB device.
+   */
+  const char* getVendorName (unsigned index) const throw ();
+
+  /**
+   * @author Suat Gedikli
+   * @brief method to retrieve the product name of the USB device without creating it.
+   * @param[in] index the index of the device in the device list.
+   * @return the product name of the USB device.
+   */
+  const char* getProductName (unsigned index) const throw ();
+
+  /**
+   * @author Suat Gedikli
+   * @brief method to retrieve the vendor id of the USB device without creating it.
+   * @param[in] index the index of the device in the device list.
+   * @return the vendor id of the USB device.
+   */
+  unsigned short getVendorID (unsigned index) const throw ();
+
+  /**
+   * @author Suat Gedikli
+   * @brief method to retrieve the product id of the USB device without creating it.
+   * @param[in] index the index of the device in the device list.
+   * @return the product id of the USB device.
+   */
+  unsigned short getProductID (unsigned index) const throw ();
+
+  /**
+   * @author Suat Gedikli
+   * @brief method to retrieve the bus id of the USB device without creating it.
+   * @param[in] index the index of the device in the device list.
+   * @return the bus id of the USB device.
+   */
+  unsigned char  getBus (unsigned index) const throw ();
+
+  /**
+   * @author Suat Gedikli
+   * @brief method to retrieve the vaddress of the USB device without creating it.
+   * @param[in] index the index of the device in the device list.
+   * @return the address of the USB device.
+   */
+  unsigned char  getAddress (unsigned index) const throw ();
+
+  /**
+   * @author Suat Gedikli
+   * @brief stops all streams from all devices.
+   */
+  void stopAll ();
+
+  /**
+   * @author Suat Gedikli
+   * @brief decomposes the connection string into vendor id and product id.
+   * @param[in] connection_string the string containing the connection information
+   * @param[out] vendorId the vendor id
+   * @param[out] productId the product id
+   */
+  static void
+  getDeviceType (const std::string& connection_string, unsigned short& vendorId, unsigned short& productId);
+protected:
+
+  struct PCL_EXPORTS DeviceContext
+  {
+    DeviceContext (const xn::NodeInfo& device_node, xn::NodeInfo* image_node, xn::NodeInfo* depth_node, xn::NodeInfo * ir_node);
+    DeviceContext (const xn::NodeInfo & device_node);
+    DeviceContext (const DeviceContext&);
+    xn::NodeInfo device_node;
+    std::shared_ptr<xn::NodeInfo> image_node;
+    std::shared_ptr<xn::NodeInfo> depth_node;
+    std::shared_ptr<xn::NodeInfo> ir_node;
+    OpenNIDevice::WeakPtr device;
```

Comment:
Why not just `pcl::weak_ptr<OpenNIDevice>`?

Reply:
Consistency with https://github.com/PointCloudLibrary/pcl/blob/8259d96aa62362f245d9a139cf3df69af39b5470/io/include/pcl/io/openni_camera/openni_device.h#L82-L83

Reply:
I'll change this as well seeing https://github.com/PointCloudLibrary/pcl/pull/3753#discussion_r402046127